Dev UI

sashakmurray opened the pull request:
Users are able to use buttons to pause the game and edit which cells are alive.

jonathan-d-zhang committed:
Add test for neighbors function

jonathan-d-zhang committed:
Change `step` function to be more testable

jonathan-d-zhang committed:
Add tests for `step` function

jonathan-d-zhang committed:
Create lookup tables for getting new point state

Removed hashset implementation.

jonathan-d-zhang committed:
Create lookup tables for getting new point state

Removed hashset implementation.

jonathan-d-zhang committed:
Implement unoptimized version for testing

jonathan-d-zhang committed:
Implement unoptimized version for testing

jonathan-d-zhang committed:
Modify python interface to work with new impl

jonathan-d-zhang committed:
Modify python interface to work with new impl

jonathan-d-zhang committed:
Fix a bug causing an index out of bounds

Caused by trying to a negative signed integer to an unsigned int

jonathan-d-zhang committed:
Fix a bug causing an index out of bounds

Caused by trying to a negative signed integer to an unsigned int

jonathan-d-zhang committed:
Fix bug that caused weird behavior

Due to code that tried to center the squares on the screen

jonathan-d-zhang committed:
Fix bug that caused weird behavior

Due to code that tried to center the squares on the screen

jonathan-d-zhang committed:
Allow cells to wrap to opposite sides

jonathan-d-zhang committed:
Allow cells to wrap to opposite sides

jonathan-d-zhang committed:
Add mapping so cells will fill the screen.

Number of cells will now scale with SQR_LEN and SCREEN_SIZE

jonathan-d-zhang committed:
Add mapping so cells will fill the screen.

Number of cells will now scale with SQR_LEN and SCREEN_SIZE

sashakmurray committed:
Update gitignore

sashakmurray committed:
Make graphics object oriented

This will make it easier to add buttons and allow the user to interact with the window.

sashakmurray committed:
Merge branch 'dev-ui' of https://github.com/SavagePastaMan/game-of-life into dev-ui

sashakmurray committed:
Make `squares` a window attribute.

Having it as a global was kinda weird

sashakmurray committed:
Update drawing of squares to only have to draw one rectangle.

Also add temporary self.paused variable

sashakmurray committed:
Move constants to constants.py

This will allow the constants to be imported to other files without circular import issues

sashakmurray committed:
Add pausing and editing functionality

Users can pause the game at any time by pressing the pause button, and resume by pressing it again.
While the game is paused, users can change which cells are alive by clicking on them.

sashakmurray committed:
Rename `ui` directory to `images`

/shrug

sashakmurray committed:
Add text to indicate if game is paused or user is editing.

sashakmurray committed:
Remove pause button

Now there is only an edit button. This will hopefully reduce confusion and make it easier to edit.

jonathan-d-zhang committed:
Fix bug caused by hard coded value in calculation

`on_mouse_press` used the literal `16` instead of `SQR_LEN`
when mapping the x and y values to indices, which caused scaling
issues when the SQR_LEN was modified.
Fixed by replacing `16` with `SQR_LEN`

sashakmurray committed:
Add exit button

This allows the user to easily exit the game by clicking the button.

sashakmurray committed:
Update __main__.py

sashakmurray committed:
Add background

This ensures that the buttons will not cover up any squares in the game.

sashakmurray committed:
Add map_index function to map coordinates to indices on the list

jonathan-d-zhang commented on src/__main__.py:
```diff
@@ -1,34 +1,132 @@
+            j = map_index(x)
+            self.squares[i][j] = not self.squares[i][j]
+
+    def draw_infobox(self, text: str) -> None:
```
could be made into a free function or static

jonathan-d-zhang reviewed (changes requested):
would also be neat to have a "tick" counter, that would show the current generation.

sashakmurray committed:
Add counter to show which generation the game is on.

sashakmurray committed:
Fix bug which caused IndexError

Previously, IndexErrors were raised when the user clicked outside of the
main screen while in edit mode.

sashakmurray committed:
Make `draw_infobox` a static method.

jonathan-d-zhang commented on src/__main__.py:
```diff
@@ -91,9 +88,11 @@ def on_mouse_press(self, x, y, _button, _modifiers):
             i = map_index(y)
             j = map_index(x)
-            self.squares[i][j] = not self.squares[i][j]
+            if i in range(LIST_LEN) and j in range(LIST_LEN):
```
```suggestion
            if 0 <= i < LIST_LEN and 0 <= j < LIST_LEN:
```

sashakmurray committed:
Did the thing

:P

sashakmurray committed:
Merge branch 'master' into dev-ui
